### PR TITLE
fix and update parts and schemas

### LINKED ISSUE #122: update parts and schemas for v0.1 release
The current parts and schemas are outdated, and v1.0/v1.1 schemas are missing.

- Update parts and schemas.
- Generate v1.1 schema, and update docs/specs/generate-validation-schema with instructions.
- Test v1.1 data with v1.1 schema.

Also, fix version1 logic:
> The version 1 fields have like a cascading validation, for example,
> 1. version1Table should always be there
> 2. version1Variable should be there if the version1Location is variables or categories
> 3. version1Category should be there if the version1Location is categories
> 
> The above logic does not apply if a part was introduced in version 2 i.e. firstReleased is 2.0

Also, make sure that version1 parts get's their table_id from version1Table:
> The best way to test this would be to,
> 1. Create a part which was created in version 1 but depreciated in version 2. Set its version1Table to a value that's not NA. This means that it won't belong to any tables in version 2.
> 2. Create a version 1 schema with it.
> 3. With your logic I believe that an incorrect schema should be generated

### PR DESCRIPTION
**Pull request by zargot**
fixes #122

- add updated parts and schemas
- add gen of v1 schemas to the schema gen tool
- add xlsx validation tool
- fix a pretty bad bug where cerb rules were being repeated
- fix allowed rule
- fix gen of v1 schemas
- fix minlength rule
- fix table attribute mapping
- fix type rule
- fix v1 catsets
- fix v1 validation
- fix doc gen error due to strip of partID that isn't a string, but an int??

Trace for later investigation of doc gen error:
https://github.com/Big-Life-Lab/PHES-ODM-Validation/actions/runs/4088485603/jobs/7050208854

### CODE REVIEW

File: src/odm_validation/part_tables.py [hunk 1]
```diff
@@ -372,59 +400,106 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
...
-        if not is_compatible(attr, ODM_VERSION):
-            for id_v1 in _parse_version1Field(attr, V1_TABLE):
-                id_v2 = table_ids_v1_v2[id_v1]
-                attr_table_ids.append(id_v2)
-        else:
-            attr_table_ids.append(_get_table_id(attr))
-        assert len(attr_table_ids) > 0
-        for id in attr_table_ids:
-            table_attrs[id].append(attr)
+    def _is_cat_v1(p: Part) -> bool:
```
Comment (yulric) [inline src/odm_validation/part_tables.py:427]: A version 1 category should also have a version 1 variable

**Review by yulric — CHANGES_REQUESTED** [review]
1. One of your commits was to remove duplicate entries in the forbidden and allowed fields in the schema. Can you explain why that was hapenning?
2. We should figure out a way to not hardcode the versions for which we need to generate validation schemas. Ideally we would get those from the firstReleased and lastUpdated columns in the dictionary
3. We need to handle multiple values (they're seperated by ;) in the the version 1 columns (version1Table, version1Variable, and version1Category)
4. I'm not sure how the tests for version 2 schemas are passing since the firstReleased and lastUpdated are set to v1 in the code by default? And we're not adding those columns in the test parts 
5. In the tests for invalid_category the following changes should be made
    1. version 1 category parts should have their status "depreciated"
    2. Their firstReleased and lastUpdated should be v1
    3. Their catSetID should be NA
6. We should refactor how the categories are being modelled for version 1 categorical attributes.

Only 3, 4, and 5 are urgent. Everything else can wait.

**Comment by zargot** [comment]
> 1. One of your commits was to remove duplicate entries in the forbidden and allowed fields in the schema. Can you explain why that was hapenning?

The forbidden rule duplication stopped happening for some reason, so i reverted that commit.
Allowed rule duplication seems to come from wrong handling of v1 categories due to them belonging to multiple tables. I've added a FIXME for that.

> 2. We should figure out a way to not hardcode the versions for which we need to generate validation schemas. Ideally we would get those from the firstReleased and lastUpdated columns in the dictionary

I'm assuming this is something we will do in a separate PR. Can you make an issue for it?

> 3. We need to handle multiple values (they're seperated by ;) in the the version 1 columns (version1Table, version1Variable, and version1Category)

Where was this not handled? I can't remember.

> 4. I'm not sure how the tests for version 2 schemas are passing since the firstReleased and lastUpdated are set to v1 in the code by default? And we're not adding those columns in the test parts

This seems to happen due to the logic in `is_compatible` which currently only cares about that the part is active, as long as the schema version is greater than the part version. I'll see if I can fix that now.

> 5. In the tests for invalid_category the following changes should be made
>    
>    1. version 1 category parts should have their status "depreciated"
>    2. Their firstReleased and lastUpdated should be v1
>    3. Their catSetID should be NA

TODO

> 6. We should refactor how the categories are being modelled for version 1 categorical attributes.

I'm assuming you're talking about how the code structures the category data. I agree that we have to improve it.

File: src/odm_validation/schemas.py [hunk 2]
```diff
@@ -22,7 +24,7 @@ def init_attr_schema(attr_id: str, rule_id: str, cerb_rules: dict,
                      meta: Meta):
     # ['meta'] is a list with one entry per rule.
     # ['meta']['meta'] is a `Meta` with one `MetaEntry` per part.
-    inner = cerb_rules
+    inner = deepcopy(cerb_rules)
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/schemas.py:28]: done

File: tools/generate_schemas.py [hunk 3]
```diff
@@ -15,26 +17,41 @@
...
 
-# setup default logger (to file)
+# setup logging
 log_dir = normpath(join(root_dir, 'logs'))
 log_file = join(log_dir, 'generate_schemas.log')
 logging.basicConfig(
-    level=logging.DEBUG,
+    level=logging.WARNING,
     handlers=[
+        logging.StreamHandler(),
```
[the remark this answers is not in the document]
Comment (zargot) [reply tools/generate_schemas.py:31]: done

File: src/odm_validation/part_tables.py [hunk 4]
```diff
@@ -200,16 +209,28 @@ def _parse_version1Field(part, key) -> List[str]:
     return list(map(str.strip, raw_ids))
 
 
-def get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
+def _normalize_key(key: Optional[str]) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/part_tables.py:213]: done

File: src/odm_validation/part_tables.py [hunk 5]
```diff
@@ -219,13 +240,18 @@ def get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
...
     except KeyError:
         return
-    if len(list(filter(lambda id: id and id != '', ids))) == 0:
+    valid_ids = list(filter(lambda id: id and id != '', ids))
+    if len(valid_ids) == 0:
         return
     return ids
 
 
+def get_part_active(part) -> bool:
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/part_tables.py:250]: done. I went with `is_active(part)`.

File: src/odm_validation/part_tables.py [hunk 6]
```diff
@@ -200,16 +209,28 @@ def _parse_version1Field(part, key) -> List[str]:
...
 
-def get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
+def _normalize_key(key: Optional[str]) -> Optional[str]:
+    """Returns `key` with the first char in lower case."""
+    if not key:
+        return
+    return key[0].lower() + key[1:]
+
+
+def _get_mappings(part: dict, version: Version) -> Optional[List[PartId]]:
```
[the remark this answers is not in the document]
Comment (zargot) [reply src/odm_validation/part_tables.py:220]: here we need to pass `version` to determine if there is a mapping (and should be a mapping) from ODM_VERSION to `version`. If I rename it to `_get_v1_ids`, etc.,  then it doesn't make sense to pass `version` anymore, as this is supposed to determine if we are getting v1 ids or not, so I left the name the same and added a docstring instead. Is that ok?

**Comment by zargot** [comment]
I'm waiting with any v1 category-related changes until we have cleared up how they should work. ref chat.

File: src/odm_validation/part_tables.py [hunk 7]
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
...
+    #   them via the version1-table/variable pair.
+    #
+    # XXX: Some v1 categories (like "tp24s") don't belong to any attributes, so
+    # we can only assume that it shouldn't be included in the schema. It seems
+    # like this is a trend among the "*Default" category sets.
+    if version.major == 1:
+        def is_cat_v1(p: Part) -> bool:
+            "Returns true if `p` is a v1-only category."
+            return (p.get(V1_TABLE) and
+                    p.get(V1_LOCATION) == 'variableCategories' and
```
Comment (yulric) [inline src/odm_validation/part_tables.py:471]: variableCategories should be moved to a variable

File: src/odm_validation/part_tables.py [hunk 8]
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
...
+        categories_v1: List[Part] = list(filter(is_cat_v1, parts))
+
+        attr_cats = defaultdict(list)
+        for cat in categories_v1:
+            cat_tables = set(_parse_version1Field(cat, V1_TABLE))
+            cat_vars = set(_parse_version1Field(cat, V1_VARIABLE))
+
+            # find a catset attr that matches
+            for attr_id in catset_data:
+                attr = all_parts[attr_id]
```
Comment (yulric) [inline src/odm_validation/part_tables.py:484]: Rename to `attr_part`

File: src/odm_validation/part_tables.py [hunk 9]
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
...
+                attr_tables = set(_parse_version1Field(attr, V1_TABLE))
```
Comment (yulric) [inline src/odm_validation/part_tables.py:485]: Rename to `attr_v1_tables`

File: src/odm_validation/part_tables.py [hunk 10]
```diff
@@ -372,59 +410,93 @@ def map_ids(mappings: Dict[PartId, PartId], part_ids: List[PartId],
...
+                attr_vars = set(_parse_version1Field(attr, V1_VARIABLE))
```
Comment (yulric) [inline src/odm_validation/part_tables.py:486]: Rename to `attr_v1_vars`

**Review by yulric — CHANGES_REQUESTED** [review]
1. Can we add the status field back into the test files? They should be the source of truth for tests and adding some of the columns in the code is confusing.

**Review by yulric — APPROVED** [review]

**Pull request merged**